### ISSUE TITLE
Agent's bypass TFS Proxy Server

### ISSUE THREAD

**Issue by Scott-Emberson** [Microsoft/vsts-agent#1356]
So in our company, we have multiple global sites, all with their own build servers and proxy servers, to help decrease build time. We have started to move over to vsts-agents now and I am finding that the vsts agent bypasses the TFS proxy server. 

##Setup Details

## Agent version and platform
2.124.0 x64

OS of the machine running the agent
Windows Server 2016

## VSTS type and version
On-Prem TFS

If On-Prem TFS, which release, 2017.1

**Comment by bryanmacfarlane** [comment]
What user is the agent running as?  Is the proxy configured for that user?  I would log on as that user and configure the proxy, ensure it's working from the cli.

**Comment by TingluoHuang** [comment]
@Scott-Emberson can you share your agent diag log? or build log that indicate it bypass your http proxy server?  For `TFS proxy server`, do you mean the TFVC proxy server TFS provided for download TFVC source content?

**Comment by TingluoHuang** [comment]
@Scott-Emberson if you are using `TFS proxy server` for your tfvc content, you want to run `%AgentRoot%\externals\vstsom\tf.exe settings proxy /configure http://tfvcproxyserver:8081`

**Comment by Scott-Emberson** [comment]
@bryanmacfarlane 
The agents connect to TFS as local accounts and run as a windows service as Network Service
@TingluoHuang 
Apologies yes I am talking about TFVC proxy server (I have updated my original issue)
I have run the command you suggested and it completes fine but I see no change. 
I can confirm the proxy is working correct due to testing it with Visual Studio

**Comment by bryanmacfarlane** [comment]
@Scott-Emberson - as a troubleshooting step (at a minimum), try to configure an agent running as a user account (agents can be sxs on the same machine, use a different name).  Log onto the machine as that user and configure the tf proxy.  From the cli, ensure it's using the proxy.  Run builds using that agent.

**Comment by TingluoHuang** [comment]
@Scott-Emberson base on source code of tf.exe, the tfvc proxy setting is stored at User Registry, so please try run the agent as the same user you run `tf settings proxy /configure`

**Comment by bryanmacfarlane** [comment]
@Scott-Emberson - any update on the troubleshooting steps above?

**Comment by Scott-Emberson** [comment]
@bryanmacfarlane - So setting the build agent to run as the local user and logging in and setting the TFVC proxy address against the build agent and the user's registry. it works fine. I have then also run the same commands against the build agent running as Network Service and this fails to get from the TFVC proxy server.

**Comment by TingluoHuang** [comment]
@Scott-Emberson did you run `tf settings proxy /configure` as network service?

**Comment by Scott-Emberson** [comment]
@TingluoHuang I have now, I ran psexec -i -u "nt authority\network service" cmd.exe and then run both of the commands one against the agent and the other for the user registry

**Comment by Scott-Emberson** [comment]
@TingluoHuang - I have done some more testing on this it looks like even if you set the TFVC proxy server against the agent (%AgentRoot%\externals\vstsom\tf.exe settings proxy /configure http://tfvcproxyserver:8081) doesn't work. it only works if you set it for the user's registry.

**Comment by Scott-Emberson** [comment]
@TingluoHuang - I have finished my testing now. It looks like to make the agent's use the TFVC Proxy Server you have to do two steps.
1. Set the proxy for the user's registry
2. Set the proxy for every agent.
Doing this does make the agent's call the TFVC first before going to the main TFS Server.

**Comment by TingluoHuang** [comment]
@Scott-Emberson which user's registry? Network Service?
what i initially thought is your run `tf settings proxy /configure` as your agent's run as user, in your case which is Network Service.
is this the approach you are taking and make it work?

**Comment by Scott-Emberson** [comment]
@TingluoHuang Yes so I have to set the user's registry value for the network service and then run the `tf settings proxy /configure` against each build agent while running the CMD as Network Service.

Steps I ran:

1. Downloaded psexec from Windows Sysinternals
2. Ran CMD as Administrator, pointed it at the location of psexec
3. Ran the following command `psexec -i -u "nt authority\network service" cmd.exe`
4. Then run the command `C:\Program Files (x86)\Microsoft Visual Studio 14.0\Common7\IDE\tf.exe proxy /configure http://tfvcproxyserver:8081`
5. Ran the following command `%AgentRoot%\externals\vstsom\tf.exe settings proxy /configure http://tfvcproxyserver:8081` on every build agent.

**Comment by TingluoHuang** [comment]
@Scott-Emberson cool, thanks for sharing detail steps. 
I am wonder whether step 4 is not necessary, since any version of tf.exe should use the same registry entry. 
Instead of using `psexec`, i think you can also use the agent itself to run `tf.exe settings proxy /configure`, add a CommandLine task to your build definition to run `tf.exe settings proxy /configure`, then queue build for that definition to all agents.
Since the agent is running as NetworkService, you build will run as NetworkService, so when the build finish, your `tfvc` proxy should be configured.

**Comment by Scott-Emberson** [comment]
@TingluoHuang So it tried just running set 3 & 4, didn't work. Tried setup 3 & 5 didn't work. I could only get it to run by doing all the steps

**Comment by TingluoHuang** [comment]
@Scott-Emberson thanks for all your testing, i will update the official doc with what you discovered

**Comment by TingluoHuang** [comment]
@Scott-Emberson after read through the code, i find a way more easy way to configure the agent to use tfvc proxy server.
`Setx TFSPROXY=http://tfvcproxy:8081` then restart the agent service.

**Comment by Ajmal18** [comment]
Hi,
I have an issue with performance of GetSource when TFS Porxy is configured.
I can see vNext agent is using our proxy server (in resource monitor i can see TF.exe is using our tfs proxy in network activity) but in this case it download source.

**Problem Detail:**
Get source build step takes around 10-12 hours using proxy in vNext builds but same get operation takes only 10-15min in XAML builds.
Our main TFS application server is in Germany and we have configured local TFS proxy in India to download sources quickly.
even i try to get source by running tf get . /r /force command , it shows expected performance (i.e. it download all source in 15min)
any hint help would be sueful for us
Thank you
regards

**Comment by TingluoHuang** [comment]
@Ajmal18 you might want to post this question to https://developercommunity.visualstudio.com/spaces/8/index.html since you are able to confirm the agent has properly setup TF.exe and let it talk to your tfs proxy.
we need the owner of tf.exe to help us understand why it is slow. :)

**Comment by Ajmal18** [comment]
@TingluoHuang thank you , i will:)
